saw-script/src/SAWScript/Typechecker.hs

```diff
@@ -1041,7 +1041,10 @@ inferExpr (ln, expr) = case expr of
       ctx <- getFreshTyVar pos
       tyResult <- getFreshTyVar pos
       let ty = tBlock (PosInferred InfTerm pos) ctx tyResult
+      saveVarEnv <- gets varEnv
+      saveTyEnv <- gets tyEnv
       body' <- inferBlock ln pos ctx ty body
+      modify (\rw -> rw { varEnv = saveVarEnv, tyEnv = saveTyEnv })
```

Comment:
Not necessarily something that needs to happen here, but it might be worth creating combinators that capture these patterns of block-scoped actions, which occur in several parts of the code in this module. Something like:

```hs
withScopedVarEnv :: TI a -> TI a
withScopedVarEnv action = do
  saveVarEnv <- gets varEnv
  res <- action
  modify (\rw -> rw { varEnv = saveVarEnv })
  pure res

withScopedTyEnv :: TI a -> TI a
withScopedTyEnv action = do
  saveTyEnv <- gets tyEnv
  res <- action
  modify (\rw -> rw { tyEnv = saveTyEnv })
  pure res
```

Then you can just call `withScopedVarEnv $ withScopedTyEnv $ inferBlock ln pos ctx ty body`, without needing to muck about with explicit environment management.